### PR TITLE
net/sslh: Initial plugin version

### PR DESCRIPTION
**Pull request by agh1467**
This is to address the discussion about a plugin for sslh and resolves https://github.com/opnsense/plugins/issues/1630. This is such a plugin.

* Includes setting listen addresses, and protocol targets.
* Includes some other advanced settings
* Service start/stop/restart control

### CODE REVIEW

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (fabianfrz) [inline net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:40]: since you are on FreeBSD here, this cannot work. The firewall is pf, not iptables / nftables

**Review by fabianfrz — CHANGES_REQUESTED** [review]

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (agh1467) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:40]: I'll remove it.

File: net/sslh/Makefile
Comment (fichtner) [inline net/sslh/Makefile:3]:
```suggestion
PLUGIN_VERSION=		0.1
PLUGIN_DEVEL=	yes
```

We do have separate versioning for software and actual plugin since both a different entities that can receive updates independently.

File: net/sslh/Makefile
Comment (fichtner) [inline net/sslh/Makefile:6]: I've dropped a mail into this direction. Unwilling to use it if you cannot be reached.

File: net/sslh/pkg-desc
Comment (fichtner) [inline net/sslh/pkg-desc:1]: It's "pkg-descr" and can you merge the changelog into it? Usually we also use the text from the FreeBSD port description since people may not know what "sslh" is otherwise (you can see this from the GUI before install with the details button in the plugins view).

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (fichtner) [inline net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:4]: Same here for mail and name... no good deed may go unpunished :)

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (fichtner) [inline net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:47]:
```suggestion
    $conf = parse_ini_file('/usr/local/opnsense/service/conf/actions.d/actions_sslh.conf', true);
```

file should be here

Comment (fichtner) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:47]: Oh wait you mean to add a generic configuration... the question would be is this used somehwere else? That directory may not be the best place for it since no relation to configd then. Apologies for doing a linear review from file to file...

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (fichtner) [inline net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:70]: All things considered the question is if it's worth it. The pid file is usually also defined in the rc.d file of the service that comes with the port. Softcoding it in a second place seems less favourable than hardcoding it once here.

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/SettingsController.php
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/SettingsController.php:4]: you might want to add your own, I don't see a reason to split hairs on copyright for a file with 2 lines of "code"

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:9]:
```suggestion
        <help>This will enable the SSLH service.</help>
```

where possible this short inline style would be the best approach for translation work and maintenance

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:15]: further problems for translations: we don't add strings that offer markup since it could break the html/js, is prone to XSS. Markup and layout inside translations remain a huge problem with no easy solution: https://github.com/opnsense/lang/issues/1

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:26]: does the tokenizer support hints in this way that are easily distinguishable for the user? Usually we provide a default that works like localhost:443 and since the field can't/shouldn't be empty that is all that is required to get started

File: net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Menu/Menu.xml
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Menu/Menu.xml:3]: Even if it looks silly usually we opt for "Sslh" although that looks strange...  a bit like Dnsmasq.

```suggestion
        <SSLH cssClass="fa fa-lock fa-fw" url="/ui/sslh/settings">
```

visible name is already set via xml node.

File: net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml:3]:
```suggestion
    <version>0.0.1</version>
```

model versioning should also be different from software version. can be linked to plugin version, but most of the time it isn't and it doesn't matter as long as version is increased when necessary (changed defaults, new required values, removed items).

File: net/sslh/src/opnsense/service/conf/actions.d/actions_sslh.conf
Comment (fichtner) [inline net/sslh/src/opnsense/service/conf/actions.d/actions_sslh.conf:13]: "one"start etc. suggests you don't have rc.conf wired up and you should do that via templates. it makes autostart on boot possible...

File: net/sslh/src/opnsense/service/conf/actions.d/actions_sslh.conf
Comment (fichtner) [inline net/sslh/src/opnsense/service/conf/actions.d/actions_sslh.conf:20]: || exit 0 shouldn't be used for anything but status where unfortunately we get an error when the service is not running which is not really an error if we stopped it / don't have it configured.

File: net/sslh/src/opnsense/service/conf/sslh.conf
Comment (fichtner) [inline net/sslh/src/opnsense/service/conf/sslh.conf:1]: looking at this file for the first time the idea is good, but I doubt softcoding these never-changing values provides much benefit.

**Review by fichtner — COMMENTED** [review]
thanks so far. I don't see a reason why we can't add this soon after addressing review. My main concerns are maintainability for the translations. I know that other code doesn't do this better but anything other than that "other" code is good progress. :)

File: net/sslh/Makefile
Comment (fichtner) [reply net/sslh/Makefile:6]: nope, message bounced

Comment (fabianfrz) [reply net/sslh/Makefile:6]: @fichtner This is an "email address" that is used by github to connect commits to a specific user when the user desires not to expose his / her email address in commits. I guess there is no forwarding by github.

File: net/sslh/Makefile
Comment (agh1467) [reply net/sslh/Makefile]: I've changed it as you've put here.

I thought about the versioning for a while with respect to dnscrypt-proxy as well. That application has phased out configuration settings, renamed settings, and introduced new settings throughout the versions. Keeping the plugin version in lock step with the application version seemed like the most practical approach since it makes it clear to the user which version of the application is either being installed, or which version the plugin is compatible with, and eliminates the need to have a separate versioning schema for the plugin. Utilizing PLUGIN_REVISION would allow a simpler incremental versioning which could reset with each major version or increment forever, and there wouldn't be any trouble with defining a versioning schema.

Is there any guidance on this? A quick search only netted me the Hello World plugin example which doesn't detail versioning. With some guidance I think I could wrap my head around when each component should be versioned and which component gets versioned under which conditions.

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (agh1467) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:47]: Yup, that's exactly it. The intent here is to define pseudo-"super globals." I provide a broader explanation on the comment of the conf file itself.

An example of how this could be useful is, this file can be copied wholesale, change the file name, and only have to change this line, and the function name to adapt to a new service. I'd guess, it could be reduced further and eliminate the need for this file at all if this information can just be derived from the conf file. For simple services the three actions are probably enough.

Less changing values after copying/pasting the code to stand up a new plugin.

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (agh1467) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:70]: I happened across this by accident since I copied this file from the dnscrypt-proxy plugin which uses `pid` as the key name for this value.
https://github.com/opnsense/plugins/blob/master/dns/dnscrypt-proxy/src/etc/inc/plugins.inc.d/dnscryptproxy.inc#L51

Since for dnscrypt-proxy, it happens that the service name and the process name actually match, everything works without issue. However, with sslh, the process names are one of either `sslh-fork` or `sslh-select` which doesn't match the 'name' value in the array, and thus it's not able to find the process by name. It only shows the service as stopped in the status column. So, for this case, the `pidfile` value is how `is_process_running()` is able to find the process and properly detect if it's running or not.

For this case, the pid file is a command line option/configuration option, and is explicitly defined in the configuration file. It's also defined in the conf for reference here. I also looked in the `/usr/local/etc/rc.d/sslh` but for sslh it didn't look like one was defined as I've seen in other services.

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (agh1467) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml]: This makes me cry a little bit. I did this because it looks nice in the XML, and presents well on the page. I understand though. My first question, to which the answer is: You can't, is how do I do paragraphs?

I'd like to better understand this challenge, as I'd like to help if I can. I've seen the usage of gettext and _lang, but I don't see either being used here. I don't understand exactly how the translations work, any help there would be welcome.

I chose this style for the help messages because I've come across many help messages for settings in OPNsense where I could only assume that a particular setting is related to a setting in a service or application. Having this information available for anyone who is familiar with the application would be really helpful especially when it comes to looking up those settings in application documentation. OpenVPN was one standout experience I had with this because many of the settings are short-handed, but long in the UI, so it makes cross referencing difficult.

Not for here to discuss, but an idea which may help with the translation issues, utilizing pre defined styles
```xml
<help>
  <header>-p, --listen address:port</header>
  <body>Paragraph one.</body>
  <body>Paragraph two.</body>
</help>
```
Then do styling using the layout_partials. Would have to figure out which styles are useful, but that would be the basic idea. Can't do it with the current `parseFormNode()` but I've written one that can.

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (agh1467) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:14]: It does, looks like this:
image

I chose the hint method since the documentation explains thusly:
```
Interface and port on which to listen, e.g. foobar:443, where
foobar is the name of an interface (typically the IP address on
which the Internet connection ends up).
```

Which would be complicated to accomplish via the model defined default. In my cursory test using `localhost:443` doesn't work.

I'm not married to the hint if a proper default could be defined in the model.

File: net/sslh/src/opnsense/service/conf/sslh.conf
Comment (agh1467) [reply net/sslh/src/opnsense/service/conf/sslh.conf:1]: Understandable with a small plugin like this as an example, it could be used more extensively with larger plugins.

These settings can then be used throughout the plugin, in PHP( and thus volt) via parse_ini_file, and in python via configparse module (still exploring ideas on getting them into the Jinja template environment). With the outcome being code which is more portable, instead of having these values statically defined throughout the entire plugin. Hopefully it would be less work for someone else tomorrow.

I wasn't sure about the location of the conf file since that directory is actually for configd and the conf sits alongside `configd.conf`, but shouldn't clobber anything if it were to also be used by plugins too. Maybe another location is more suitable?

**Comment by agh1467** [comment]
Thanks all for the review and comments! I resolved the comments that were straight forward and changes were made. I left the others open which I thought there might be more to say.

**Review by fabianfrz — APPROVED** [review]

File: net/sslh/Makefile
Comment (fichtner) [reply net/sslh/Makefile]: The problem is you can't be there for any update of dnscrypt-proxy which will make these versions differ and having a separate versioning for both plugin and dependent software makes it easier to see what people talk about. Sometimes people forget to name the revision, too. That being said if you have more than one software dependency how would you decide which software dependency gets to be the authoritative version number? ;)

Besides, revision is used as a means to not tamper with the version number, but you can't increment it "forever" as it makes no sense for hotfixing annotation. Yes, the idea comes from FreeBSD ports where you package a software and can't control the version number, but it's still clearer than not having a revision. The difference at the end of the day is 1.1.1 vs 1.1_1.

There's no document for the versioning aspects in our case. The unwritten rules are 0.x for development, 1.x and up for releases. Usually 2.x denotes a rewrite and some maintainers prefer 1.x.y and when we make quick bugfixes / hotfixes we use the revision instead to sort of denote that the fix may or may not be handled by the maintainer. More often than not, however, hotfixes are proper versions served by maintainers (hello @fraenki 😄). Revisions are cleared when the version changes. 

One more thing to keep in mind here is that the makefile denotes the versioning and upgrade behaviour. We don't have any automatic means to update a plugin if this information wasn't modified (either version or revision in case version should stay the same). A year ago we did not even have a commit hash in the packaged plugin. But more on this in another comment here.

File: net/sslh/Makefile
Comment (fichtner) [reply net/sslh/Makefile]: @agh1467 thanks, highly appreciated

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (fichtner) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:47]: Ok in any case the file needs to be moved. A more fitting location would be either `/usr/local/etc/inc/plugins.inc.d/sshl/sshl.conf` or `/usr/local/opnsense/data/sshl/sshl.conf`.

That being said we already have a plugin-bound meta information file that would be most fitting for these types of values, e.g.:

```
# cat /usr/local/opnsense/version/ddclient 
{
    "product_abi": "22.1",
    "product_arch": "amd64",
    "product_email": "ad@opnsense.org",
    "product_flavour": "LibreSSL",
    "product_hash": "be8192b82",
    "product_id": "os-ddclient-devel",
    "product_name": "ddclient",
    "product_version": "0.1",
    "product_website": "https://opnsense.org/"
}
```

I don't mind extending the opnsense-version utility to read manual values from it as long as we can avoid too much clutter. What I don't think is a great idea is to read this file on each file load of a model or plugin invoke, however. And I think @AdSchellevis will agree. For configuration purposes it's fine but for runtime always bouncing through the file or shell command is a performance killer.

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (fichtner) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:70]: It seems that is a bug worth fixing?

```
# git grep "'pid'"
dns/dnscrypt-proxy/src/etc/inc/plugins.inc.d/dnscryptproxy.inc:        'pid' => '/var/run/dnscrypt-proxy.pid'
net-mgmt/nrpe/src/etc/inc/plugins.inc.d/nrpe.inc:        'pid' => '/var/run/nrpe.pid'
net/ntopng/src/etc/inc/plugins.inc.d/ntopng.inc:        'pid' => '/var/run/ntopng/ntopng.pid'
```

Well, the 'name' value should match the process name (binary), at least that what is being used if no PID file is given. In the example you give it seems perfectly fine except for the key not being proper. Especially for plugins.inc/d content the file is always bound to be adapted by plugin maintainer.

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (fichtner) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml]: > You can't, is how do I do paragraphs?

You can, but we won't translate it for the mentioned reasons.

> I'd like to better understand this challenge, as I'd like to help if I can.

Model strings are dynamically generated from XML content via https://github.com/opnsense/lang/blob/master/Scripts/plugins/models.py and automatically redirected through gettext() upon form generation. This is more of an oddity of how gettext works and does not work.

> I chose this style for the help messages because I've come across many help messages for settings in OPNsense where I could only assume that a particular setting is related to a setting in a service or application. 

Essentially you are writing inline documentation in a restricted scope (time and space). Most of elaborate documentation should be on docs.git and it's actually better to have one document for each plugin as we try with the ones that we as a core team introduce nowadays. It's not a bad thing, but having a shorter help in the GUI and a full documentation indexable by search engines has multiple advantages already: less strain on translations, simpler GUI, more time for users to do what they want to do while working with the GUI, full text search from search engine and documentation, examples and tutorials in documentation, more visibility for non-OPNsense users on the topic of the plugin.

> Not for here to discuss, but an idea which may help with the translation issues, utilizing pre defined styles

That looks like what we had in mind. The downside (I'm sorry to play devil's advocate) is that while the string are now clean and well-formed we essentially end up with splitting a block of translation that belongs together into separate strings that may or may not be translated, may or may not be translated by the same persons at the same time. Context could get lost. That's why for the most part we would prefer the documentation over this.

It would be nice if you could dump your idea in that lang.git issue for further discussion. Thanks :)

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (fichtner) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml:14]: The hint is fine, but I would suggest avoiding the second entry altogether to not confuse users how tokenizer input should look like. It will lead to questions later.

localhost:443 probably will not work because the web GUI listens on it.

But the real question here is how the model is designed: do you need this entry in the configuration? If not what is the default or if you require it from the model you will force an initial model migration to fail if you don't set a default value which masks the hint you put there to begin with. :) hints are mostly useful for optional fields or when you properly embed all the constraints in the model (if enabled the hostname cannot be empty).

File: net/sslh/src/opnsense/service/conf/sslh.conf
Comment (fichtner) [reply net/sslh/src/opnsense/service/conf/sslh.conf:1]: I made two suggestions about the location elsewhere. As said before the main concern now is reading the file for read operations of the GUI... that could slow things down eventually if more plugins use this approach.

**Comment by agh1467** [comment]
FYI, I'm planning on revisiting this soon, and re-working parts to address the topics brought up.

File: net/sslh/Makefile
Comment (agh1467) [reply net/sslh/Makefile]: Thanks for the explanation on versioning. Is there a place in the OPNsense docs this info could go? I could give a shot at writing something up. Maybe in the hello world example? https://docs.opnsense.org/development/examples/helloworld.html

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (agh1467) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc]: Since the practicality for this isn't represented in this small plugin I've decided to remove it and replace with hard coded strings.

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (agh1467) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc]: I'll take a look at the other plugins and submit PRs for those as well.

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (agh1467) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml]: I recently dug into the opnsense/lang repo while researching something else, and I believe I've gained a better understanding of the challenge. I've changed them to just a plain text description.

You mentioned the docs git, and I hadn't considered writing plugin docs there. I'm thinking the README I've put together will be a good base for such documentation. I'm guessing this is where it would live? https://github.com/opnsense/docs/tree/master/source/manual/how-tos

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml
Comment (agh1467) [reply net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/forms/settings.xml]: The combination of required field without a default and how it would affect model migration was not something I was aware of. I've set a sane/functional default in the model, and removed the hint.

File: net/sslh/Makefile
Comment (fichtner) [inline net/sslh/Makefile:4]: most plugins are frontends. the opportunity here is to describe the software being configured in a nutshell :)

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (fichtner) [inline net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:4]:
```suggestion
    Copyright (C) 2022 NAMEORGITHUBHANDLE <agh1467@protonmail.com>
```

something like this would be nice, ideally we encourage natural names but I can understand not everyone wants that. first name is ok too

File: net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/Api/SettingsController.php
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/controllers/OPNsense/Sslh/Api/SettingsController.php:131]: can leave this in but doesn't look like functional addition. less is often more

File: net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml
Comment (fichtner) [inline net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml:11]: validation message here but no special validation of individual values?

Comment (fichtner) [reply net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml:11]: localhots:443 is probably already bound to web GUI btw

File: net/sslh/src/opnsense/service/templates/OPNsense/Sslh/sslh.jinja
Comment (fichtner) [inline net/sslh/src/opnsense/service/templates/OPNsense/Sslh/sslh.jinja:5]: could be reduced, but not required

there's an "empty" helper that would  make sure the config array is populated before access

**Review by fichtner — APPROVED** [review]
looks good, minor comments for release version progress

**Comment by fichtner** [comment]
@agh1467 if you ping again I will merge, up to you if you want to change something now or merge first

File: net/sslh/Makefile
Comment (agh1467) [reply net/sslh/Makefile:4]: I'll update the comment to describe the sslh service.

File: net/sslh/src/etc/inc/plugins.inc.d/sslh.inc
Comment (agh1467) [reply net/sslh/src/etc/inc/plugins.inc.d/sslh.inc:4]: I'll update to have my github handle.

File: net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml
Comment (agh1467) [reply net/sslh/src/opnsense/mvc/app/models/OPNsense/Sslh/Settings.xml:11]: For this, I considered the options that I'm aware of and can understand, but I'm having a challenge understanding the best approach. I could use something like a regular expression, but the fact that this supports a hostname and IP address will make this complicated to do in single mask.

I thought about a custom FieldType with a new validator using the CallbackValidator, the challenges would be similar, but I could at least split each CSV and explode on the colon (only reliable for IPv4), and evaluate each half of each entry separately. I feel that's still re-inventing the wheel when there exists HostValidator, and PortField already has a validator (though hard coded in the FieldType itself).

I think an option might be to re-factor the field into an ArrayField, and specify the hostname/IP, and port separately and validate each according to each field type. However, I'd prefer not to do that as it would require building the interface(s) to support the user input. I wasn't planning on going down that route until the next version of the plugin.

I'm not strong in the validators area, and I left this alone due to the complexity of specifying a single mask to handle the entire thing. Let me know how this could be approached in a reasonable manner.

File: net/sslh/src/opnsense/service/templates/OPNsense/Sslh/sslh.jinja
Comment (agh1467) [reply net/sslh/src/opnsense/service/templates/OPNsense/Sslh/sslh.jinja:5]: Thanks, I'll do that. I always forget about those helper functions.

**Comment by coatmaker618** [comment]
Super interested 3rd party here who runs OPNSense & REALLY wants this feature.  What is the state of this?  If I'm reading/understanding this correctly, this change was approved but is not in the current build & is awaiting minor comments?

As an (overly) excited 3rd party, having this capability in OPNSense would be huge and I'm super thankful that it seems to be so close!  I'm just making sure it's not lost just before the finish line!

**Comment by haarp** [comment]
I know I'm a few days late to the party, but regarding the removal of transparent mode: The officical docs have a section for FreeBSD [here](https://github.com/yrutschle/sslh/blob/master/doc/tproxy.md#freebsd), so it's not entirely impossible. Unfortunately it involves ipfw instead of pf. Maybe it can be added on top of pf or adapted.

**Comment by agh1467** [comment]
@haarp Thanks for citing that. I'll have to see if there is an equivalent functionality, and translate it accordingly. It could be added back in then.

**Review by thn80 — COMMENTED** [review]
You wrote that the documentation reports this as a "Linux only" feature and that there is no provision for using this on FreeBSD.
I am wondering about this, because the documentation of sslh states for the Transparent Proxy "On Linux and **FreeBSD** you can use the --transparent option to request transparent proxying." (Source: https://github.com/yrutschle/sslh/blob/master/doc/tproxy.md).

**Pull request merged**
